fe2-android/app/src/main/java/com/github/dedis/student20_pop/model/Wallet.java

```diff
@@ -0,0 +1,341 @@
+package com.github.dedis.student20_pop.model;
+
+import android.util.Log;
+import androidx.core.util.Pair;
+import io.github.novacrypto.bip39.MnemonicGenerator;
+import io.github.novacrypto.bip39.MnemonicValidator;
+import io.github.novacrypto.bip39.SeedCalculator;
+import io.github.novacrypto.bip39.Words;
+import io.github.novacrypto.bip39.wordlists.English;
+import java.nio.charset.Charset;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.StringJoiner;
+import javax.crypto.Mac;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+import net.i2p.crypto.eddsa.Utils;
+import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters;
+import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters;
+
+/**
+ * This class represent a wallet that will enable users to store their PoP tokens with reasonable,
+ * realistic security and usability.
+ */
+public class Wallet {
+
+  private static final String TAG = Wallet.class.getSimpleName();
+  private static final String HMAC_SHA512 = "HmacSHA512";
+  private static final int PURPOSE =  888;
+  private static final int ACCOUNT =  0;
+  private byte[] SEED;
+
+  private boolean IS_SET_UP = false;
+
+  private static final Wallet instance = new Wallet();
+  public static Wallet getInstance() {
+    return instance;
+  }
+
+  /**
+   * Class constructor, initialize the wallet with a new random seed.
+   */
+  public Wallet() {
+    SecureRandom random = new SecureRandom();
+    byte[] bytes = random.generateSeed(64);
+    SEED = bytes;
+    Log.d(TAG, "Wallet initialized with a new random seed: " + Utils.bytesToHex(SEED));
+  }
+
+  /**
+   * Method to overwrite the seed of the current wallet with a new seed.
+   *
+   * @param seed
+   */
+  public void initialize(String seed){
+    if (seed == null) {
+      throw new UnsupportedOperationException("Unable to init seed from a null param!");
+    }
+    SEED = Utils.hexToBytes(seed);
+    Log.d(TAG, "New seed initialized: " + Utils.bytesToHex(SEED));
+  }
+
+  /**
+   * Method that allow generate a different key for each path that you give.
+   *
+   * @param path a String path of the form: m/i/j/k/... where i,j,k,.. are 31-bit integer.
+   * @return a Pair<byte[], byte[]> representing the keys pair: first=private_key; second=public_key.
+   * @throws NoSuchAlgorithmException
+   * @throws InvalidKeyException
+   * @throws ShortBufferException
+   */
+  public Pair<byte[], byte[]> GenerateKeyFromPath(String path)
+      throws NoSuchAlgorithmException, InvalidKeyException, ShortBufferException {
+    if (path == null) {
+      throw new UnsupportedOperationException("Unable to find keys from a null path!");
```

Comment:
I think it would rather be IllegalArgumentException than UnsupportedOperationException

Reply:
yes you have right.